File: tests/test_factory_constructors.cpp
/*
    tests/test_factory_constructors.cpp -- tests construction from a factory function
                                           via py::init_factory()

    Copyright (c) 2017 Jason Rhinelander <jason@imaginary.ca>

    All rights reserved. Use of this source code is governed by a
    BSD-style license that can be found in the LICENSE file.
*/

#include "pybind11_tests.h"
#include "constructor_stats.h"
#include <cmath>
#include <pybind11/factory.h>

// Classes for testing python construction via C++ factory function:
// Not publically constructible, copyable, or movable:
class TestFactory1 {
    friend class TestFactoryHelper;
    TestFactory1() : value("(empty)") { print_default_created(this); }
    TestFactory1(int v) : value(std::to_string(v)) { print_created(this, value); }
    TestFactory1(std::string v) : value(std::move(v)) { print_created(this, value); }
    TestFactory1(TestFactory1 &&) = delete;
    TestFactory1(const TestFactory1 &) = delete;
    TestFactory1 &operator=(TestFactory1 &&) = delete;
    TestFactory1 &operator=(const TestFactory1 &) = delete;
public:
    std::string value;
    ~TestFactory1() { print_destroyed(this); }
};
// Non-public construction, but moveable:
class TestFactory2 {
    friend class TestFactoryHelper;
    TestFactory2() : value("(empty2)") { print_default_created(this); }
    TestFactory2(int v) : value(std::to_string(v)) { print_created(this, value); }
    TestFactory2(std::string v) : value(std::move(v)) { print_created(this, value); }
public:
    TestFactory2(TestFactory2 &&m) { value = std::move(m.value); print_move_created(this); }
    TestFactory2 &operator=(TestFactory2 &&m) { value = std::move(m.value); print_move_assigned(this); return *this; }
    std::string value;
    ~TestFactory2() { print_destroyed(this); }
};
// Mixed direct/factory construction:
class TestFactory3 {
protected:
    friend class TestFactoryHelper;
    TestFactory3() : value("(empty3)") { print_default_created(this); }
    TestFactory3(int v) : value(std::to_string(v)) { print_created(this, value); }
public:
    TestFactory3(std::string v) : value(std::move(v)) { print_created(this, value); }
    TestFactory3(TestFactory3 &&m) { value = std::move(m.value); print_move_created(this); }
    TestFactory3 &operator=(TestFactory3 &&m) { value = std::move(m.value); print_move_assigned(this); return *this; }
    std::string value;
    virtual ~TestFactory3() { print_destroyed(this); }
};
// Inheritance test
class TestFactory4 : public TestFactory3 {
public:
    TestFactory4() : TestFactory3() { print_default_created(this); }
    TestFactory4(int v) : TestFactory3(v) { print_created(this, v); }
    virtual ~TestFactory4() { print_destroyed(this); }
};
// Another class for an invalid downcast test
class TestFactory5 : public TestFactory3 {
public:
    TestFactory5(int i) : TestFactory3(i) { print_created(this, i); }
    virtual ~TestFactory5() { print_destroyed(this); }
};

struct NinetyNine {};
class TestFactory6 {
protected:
    int value;
    bool alias = false;
public:
    TestFactory6(int i) : value{i} { print_created(this, i); }
    TestFactory6(TestFactory6 &&f) { print_move_created(this); value = f.value; alias = f.alias; }
    TestFactory6(const TestFactory6 &f) { print_copy_created(this); value = f.value; alias = f.alias; }
    // Implicit conversion not supported by alias:
    TestFactory6(NinetyNine) : TestFactory6(99) {}
    virtual ~TestFactory6() { print_destroyed(this); }
    virtual int get() { return value; }
    bool has_alias() { return alias; }
};
class PyTF6 : public TestFactory6 {
public:
    PyTF6(int i) : TestFactory6(i) { alias = true; print_created(this, i); }
    // Allow implicit conversion from std::string:
    PyTF6(std::string s) : TestFactory6((int) s.size()) { alias = true; print_created(this, s); }
    PyTF6(PyTF6 &&f) : TestFactory6(std::move(f)) { print_move_created(this); }
    PyTF6(const PyTF6 &f) : TestFactory6(f) { print_copy_created(this); }
    virtual ~PyTF6() { print_destroyed(this); }
    int get() override { PYBIND11_OVERLOAD(int, TestFactory6, get, /*no args*/); }
};

// Stash leaked values here so we can clean up at the end of the test:
py::object leak1;
TestFactory3 *leak2, *leak3;
class TestFactoryHelper {
public:
    // Return via pointer:
    static TestFactory1 *construct1() { return new TestFactory1(); }
    // Holder:
    static std::unique_ptr<TestFactory1> construct1(int a) { return std::unique_ptr<TestFactory1>(new TestFactory1(a)); }
    // pointer again
    static TestFactory1 *construct1(std::string a) { return new TestFactory1(a); }

    // pointer:
    static TestFactory2 *construct2() { return new TestFactory2(); }
    // holder:
    static std::unique_ptr<TestFactory2> construct2(int a) { return std::unique_ptr<TestFactory2>(new TestFactory2(a)); }
    // by value moving:
    static TestFactory2 construct2(std::string a) { return TestFactory2(a); }

    // pointer:
    static TestFactory3 *construct3() { return new TestFactory3(); }
    // holder:
    static std::shared_ptr<TestFactory3> construct3(int a) { return std::shared_ptr<TestFactory3>(new TestFactory3(a)); }
    // by object:
    static py::object construct3(double a) {
        return py::cast(new TestFactory3((int) std::lround(a)), py::return_value_policy::take_ownership); }

    // Invalid values:
    // Multiple references:
    static py::object construct_bad3a(double v) {
        auto o = construct3(v);
        leak1 = o;
        return o;
    }
    // Unowned pointer:
    static py::object construct_bad3b(int v) {
        leak2 = new TestFactory3(v);
        return py::cast(leak2, py::return_value_policy::reference);
    }
};

test_initializer factory_constructors([](py::module &m) {

    // Define various trivial types to allow simpler overload resolution:
    py::module m_tag = m.def_submodule("tag");
#define MAKE_TAG_TYPE(Name) \
    struct Name##_tag {}; \
    py::class_<Name##_tag>(m_tag, #Name "_tag").def(py::init<>()); \
    m_tag.attr(#Name) = py::cast(Name##_tag{})
    MAKE_TAG_TYPE(pointer);
    MAKE_TAG_TYPE(unique_ptr);
    MAKE_TAG_TYPE(move);
    MAKE_TAG_TYPE(object);
    MAKE_TAG_TYPE(shared_ptr);
    MAKE_TAG_TYPE(raw_object);
    MAKE_TAG_TYPE(multiref);
    MAKE_TAG_TYPE(unowned);
    MAKE_TAG_TYPE(derived);
    MAKE_TAG_TYPE(TF4);
    MAKE_TAG_TYPE(TF5);
    MAKE_TAG_TYPE(null_ptr);
    MAKE_TAG_TYPE(base);
    MAKE_TAG_TYPE(invalid_base);
    MAKE_TAG_TYPE(alias);
    MAKE_TAG_TYPE(unaliasable);

    py::class_<TestFactory1>(m, "TestFactory1")
        .def(py::init_factory([](pointer_tag, int v) { return TestFactoryHelper::construct1(v); }))
        .def(py::init_factory([](unique_ptr_tag, std::string v) { return TestFactoryHelper::construct1(v); }))
        .def(py::init_factory([](pointer_tag) { return TestFactoryHelper::construct1(); }))
        // Takes a python function that returns the instance:
        .def(py::init_factory([](py::function f) { return f(123); }))
        // Sets a fallback python factory function (gets called if none of the above match):
        .def_static("set_ctor_fallback", [](py::function f) {
            auto tf1 = py::reinterpret_borrow<py::class_<TestFactory1>>(
                    py::module::import("pybind11_tests").attr("TestFactory1"));
            tf1.def(py::init_factory([f]() { return f(); }));
        })
        .def_readwrite("value", &TestFactory1::value)
        ;
    py::class_<TestFactory2>(m, "TestFactory2")
        .def(py::init_factory([](pointer_tag, int v) { return TestFactoryHelper::construct2(v); }))
        .def(py::init_factory([](unique_ptr_tag, std::string v) { return TestFactoryHelper::construct2(v); }))
        .def(py::init_factory([](move_tag) { return TestFactoryHelper::construct2(); }))
        .def_readwrite("value", &TestFactory2::value)
        ;
    int c = 1;
    // Stateful & reused:
    auto c4a = [c](pointer_tag, TF4_tag, int a) { return new TestFactory4(a);};
    auto c4b = [](object_tag, TF4_tag, int a) {
        return py::cast(new TestFactory4(a), py::return_value_policy::take_ownership); };

    py::class_<TestFactory3, std::shared_ptr<TestFactory3>>(m, "TestFactory3")
        .def(py::init_factory([](pointer_tag, int v) { return TestFactoryHelper::construct3(v); }))
        .def(py::init_factory([](shared_ptr_tag) { return TestFactoryHelper::construct3(); }))
        .def("__init__", [](TestFactory3 &self, std::string v) { new (&self) TestFactory3(v); }) // regular ctor
        // Stateful lambda returning py::object:
        .def(py::init_factory([c](object_tag, int v) { return TestFactoryHelper::construct3(double(v + c)); }))
        .def(py::init_factory([](raw_object_tag, double v) {
            auto o = TestFactoryHelper::construct3(v); return o.release().ptr(); }))
        .def(py::init_factory([](multiref_tag, double v) { return TestFactoryHelper::construct_bad3a(v); })) // multi-ref object
        .def(py::init_factory([](unowned_tag, int v) { return TestFactoryHelper::construct_bad3b(v); })) // unowned ptr
        // wrong type returned (should trigger static_assert failure if uncommented):
        //.def(py::init_factory([](double a, int b) { return TestFactoryHelper::construct2((int) (a + b)); }))

        // factories returning a derived type:
        .def(py::init_factory(c4a)) // derived ptr
        .def(py::init_factory(c4b)) // derived py::object: fails; object up/down-casting currently not supported
        .def(py::init_factory([](pointer_tag, TF5_tag, int a) { return new TestFactory5(a); }))
        .def(py::init_factory([](pointer_tag, TF5_tag, int a) { return new TestFactory5(a); }))
        // derived shared ptr:
        .def(py::init_factory([](shared_ptr_tag, TF4_tag, int a) { return std::make_shared<TestFactory4>(a); }))
        .def(py::init_factory([](shared_ptr_tag, TF5_tag, int a) { return std::make_shared<TestFactory5>(a); }))

        // Returns nullptr:
        .def(py::init_factory([](null_ptr_tag) { return (TestFactory3 *) nullptr; }))

        .def_readwrite("value", &TestFactory3::value)
        .def_static("cleanup_leaks", []() {
            leak1 = py::object();
            // Make sure they aren't referenced before deleting them:
            if (py::detail::get_internals().registered_instances.count(leak2) == 0)
                delete leak2;
            if (py::detail::get_internals().registered_instances.count(leak2) == 0)
                delete leak3;
        })
        ;
    py::class_<TestFactory4, TestFactory3, std::shared_ptr<TestFactory4>>(m, "TestFactory4")
        .def(py::init_factory(c4a)) // pointer
        .def(py::init_factory(c4b)) // py::object
        // Valid downcasting test:
        .def(py::init_factory([](shared_ptr_tag, base_tag, int a) {
            return std::shared_ptr<TestFactory3>(new TestFactory4(a)); }))
        .def(py::init_factory([](pointer_tag, base_tag, int a) {
            return (TestFactory3 *) new TestFactory4(a); }))
        // Invalid downcasting test:
        .def(py::init_factory([](shared_ptr_tag, invalid_base_tag, int a) {
            return std::shared_ptr<TestFactory3>(new TestFactory5(a)); }))
        .def(py::init_factory([](pointer_tag, invalid_base_tag, int a) {
            return (TestFactory3 *) new TestFactory5(a); }))
        ;

    // Doesn't need to be registered, but registering makes getting ConstructorStats easier:
    py::class_<TestFactory5, TestFactory3, std::shared_ptr<TestFactory5>>(m, "TestFactory5");

    // Alias testing
    py::class_<TestFactory6, PyTF6>(m, "TestFactory6")
        .def(py::init_factory([](int i) { return i; }))
        .def(py::init_factory([](std::string s) { return s; }))
        .def(py::init_factory([](base_tag, int i) { return TestFactory6(i); }))
        .def(py::init_factory([](alias_tag, int i) { return PyTF6(i); }))
        .def(py::init_factory([](alias_tag, pointer_tag, int i) { return new PyTF6(i); }))
        .def(py::init_factory([](base_tag, pointer_tag, int i) { return new TestFactory6(i); }))
        .def(py::init_factory([](base_tag, alias_tag, pointer_tag, int i) { return (TestFactory6 *) new PyTF6(i); }))
        .def(py::init_factory([](unaliasable_tag) { return NinetyNine(); }))

        .def("get", &TestFactory6::get)
        .def("has_alias", &TestFactory6::has_alias)

        .def_static("get_cstats", &ConstructorStats::get<TestFactory6>, py::return_value_policy::reference)
        .def_static("get_alias_cstats", &ConstructorStats::get<PyTF6>, py::return_value_policy::reference)
        ;
});

File: tests/test_factory_constructors.py
import pytest
from pybind11_tests import ConstructorStats


def test_init_factory_basic():
    """Tests py::init_factory() wrapper around various ways of returning the object"""
    from pybind11_tests import TestFactory1, TestFactory2, TestFactory3, tag

    cstats = [ConstructorStats.get(c) for c in [TestFactory1, TestFactory2, TestFactory3]]
    cstats[0].alive()  # force gc
    n_inst = ConstructorStats.detail_reg_inst()

    x1 = TestFactory1(tag.unique_ptr, 3)
    assert x1.value == "3"
    y1 = TestFactory1(tag.pointer)
    assert y1.value == "(empty)"
    z1 = TestFactory1("hi!")
    assert z1.value == "hi!"

    assert ConstructorStats.detail_reg_inst() == n_inst + 3

    x2 = TestFactory2(tag.move)
    assert x2.value == "(empty2)"
    y2 = TestFactory2(tag.pointer, 7)
    assert y2.value == "7"
    z2 = TestFactory2(tag.unique_ptr, "hi again")
    assert z2.value == "hi again"

    assert ConstructorStats.detail_reg_inst() == n_inst + 6

    x3 = TestFactory3(tag.shared_ptr)
    assert x3.value == "(empty3)"
    y3 = TestFactory3(tag.pointer, 42)
    assert y3.value == "42"
    z3 = TestFactory3("bye")
    assert z3.value == "bye"

    with pytest.raises(TypeError) as excinfo:
        TestFactory3(tag.null_ptr)
    assert (str(excinfo.value) ==
            "pybind11::init(): factory function returned nullptr")

    assert [i.alive() for i in cstats] == [3, 3, 3]
    assert ConstructorStats.detail_reg_inst() == n_inst + 9

    del x1, y2, y3, z3
    assert [i.alive() for i in cstats] == [2, 2, 1]
    assert ConstructorStats.detail_reg_inst() == n_inst + 5
    del x2, x3, y1, z1, z2
    assert [i.alive() for i in cstats] == [0, 0, 0]
    assert ConstructorStats.detail_reg_inst() == n_inst

    assert [i.values() for i in cstats] == [
        ["3", "hi!"],
        ["7", "hi again"],
        ["42", "bye"]
    ]
    assert [i.default_constructions for i in cstats] == [1, 1, 1]


def test_init_factory_casting():
    """Tests py::init_factory() wrapper with various upcasting and downcasting returns"""
    from pybind11_tests import TestFactory3, TestFactory4, TestFactory5, tag

    cstats = [ConstructorStats.get(c) for c in [TestFactory3, TestFactory4, TestFactory5]]
    cstats[0].alive()  # force gc
    n_inst = ConstructorStats.detail_reg_inst()

    # Construction from derived references:
    a = TestFactory3(tag.pointer, tag.TF4, 4)
    assert a.value == "4"
    b = TestFactory3(tag.shared_ptr, tag.TF4, 5)
    assert b.value == "5"
    c = TestFactory3(tag.pointer, tag.TF5, 6)
    assert c.value == "6"
    d = TestFactory3(tag.shared_ptr, tag.TF5, 7)
    assert d.value == "7"

    assert ConstructorStats.detail_reg_inst() == n_inst + 4

    # Construction from base references:
    e = TestFactory4(tag.pointer, tag.base, 8)
    assert e.value == "8"
    f = TestFactory4(tag.shared_ptr, tag.base, 9)
    assert f.value == "9"

    assert ConstructorStats.detail_reg_inst() == n_inst + 6
    assert [i.alive() for i in cstats] == [6, 4, 2]

    del a
    assert [i.alive() for i in cstats] == [5, 3, 2]
    assert ConstructorStats.detail_reg_inst() == n_inst + 5

    del b, c, e
    assert [i.alive() for i in cstats] == [2, 1, 1]
    assert ConstructorStats.detail_reg_inst() == n_inst + 2

    del f, d
    assert [i.alive() for i in cstats] == [0, 0, 0]
    assert ConstructorStats.detail_reg_inst() == n_inst

    # These return a base class which is *not* actually a TestFactory4 instance:
    with pytest.raises(TypeError) as excinfo:
        TestFactory4(tag.pointer, tag.invalid_base, -2)
    assert (str(excinfo.value) == "pybind11::init(): factory function failed: "
                                  "casting base pointer to instance pointer failed")
    with pytest.raises(TypeError) as excinfo:
        TestFactory4(tag.shared_ptr, tag.invalid_base, -3)
    assert (str(excinfo.value) ==
            "pybind11::init(): factory construction failed: base class shared_ptr is not "
            "a derived instance")

    assert [i.alive() for i in cstats] == [0, 0, 0]
    assert ConstructorStats.detail_reg_inst() == n_inst

    assert [i.values() for i in cstats] == [
        ["4", "5", "6", "7", "8", "9", "-2", "-3"],
        ["4", "5", "8", "9"],
        ["6", "7", "-2", "-3"]
    ]


def test_init_factory_alias():
    """Tests py::init_factory() wrapper with value conversions and alias types"""
    from pybind11_tests import TestFactory6, tag

    cstats = [TestFactory6.get_cstats(), TestFactory6.get_alias_cstats()]
    cstats[0].alive()  # force gc
    n_inst = ConstructorStats.detail_reg_inst()

    a = TestFactory6(tag.base, 1)
    assert a.get() == 1
    assert not a.has_alias()
    b = TestFactory6(tag.alias, "hi there")
    assert b.get() == 8
    assert b.has_alias()
    c = TestFactory6(tag.alias, 3)
    assert c.get() == 3
    assert c.has_alias()
    d = TestFactory6(tag.alias, tag.pointer, 4)
    assert d.get() == 4
    assert d.has_alias()
    e = TestFactory6(tag.base, tag.pointer, 5)
    assert e.get() == 5
    assert not e.has_alias()
    f = TestFactory6(tag.base, tag.alias, tag.pointer, 6)
    assert f.get() == 6
    assert f.has_alias()

    assert ConstructorStats.detail_reg_inst() == n_inst + 6
    assert [i.alive() for i in cstats] == [6, 4]

    del a, b, e
    assert [i.alive() for i in cstats] == [3, 3]
    assert ConstructorStats.detail_reg_inst() == n_inst + 3
    del f, c, d
    assert [i.alive() for i in cstats] == [0, 0]
    assert ConstructorStats.detail_reg_inst() == n_inst

    class MyTest(TestFactory6):
        def __init__(self, *args):
            TestFactory6.__init__(self, *args)

        def get(self):
            return -5 + TestFactory6.get(self)

    z = MyTest(tag.base, 123)
    assert z.get() == 118
    assert z.has_alias()
    y = MyTest(tag.alias, "why hello!")
    assert y.get() == 5
    assert y.has_alias()
    assert ConstructorStats.detail_reg_inst() == n_inst + 2
    assert [i.alive() for i in cstats] == [2, 2]
    del z, y
    assert [i.alive() for i in cstats] == [0, 0]
    assert ConstructorStats.detail_reg_inst() == n_inst

    assert [i.values() for i in cstats] == [
        ["1", "8", "3", "4", "5", "6", "123", "10"],
        ["hi there", "3", "4", "6", "move", "123", "why hello!"]
    ]


def test_init_factory_dual():
    """Tests init factory functions with dual main/alias factory functions"""
    from pybind11_tests import TestFactory7, tag

    cstats = [TestFactory7.get_cstats(), TestFactory7.get_alias_cstats()]
    cstats[0].alive()  # force gc
    n_inst = ConstructorStats.detail_reg_inst()

    class PythFactory7(TestFactory7):
        def get(self):
            return 100 + TestFactory7.get(self)

    a1 = TestFactory7(1)
    a2 = PythFactory7(2)
    assert a1.get() == 1
    assert a2.get() == 102
    assert not a1.has_alias()
    assert a2.has_alias()

    b1 = TestFactory7(tag.pointer, 3)
    b2 = PythFactory7(tag.pointer, 4)
    assert b1.get() == 3
    assert b2.get() == 104
    assert not b1.has_alias()
    assert b2.has_alias()

    c1 = TestFactory7(tag.mixed, 5)
    c2 = PythFactory7(tag.mixed, 6)
    assert c1.get() == 5
    assert c2.get() == 106
    assert not c1.has_alias()
    assert c2.has_alias()

    d1 = TestFactory7(tag.base, tag.pointer, 7)
    d2 = PythFactory7(tag.base, tag.pointer, 8)
    assert d1.get() == 7
    assert d2.get() == 108
    assert not d1.has_alias()
    assert d2.has_alias()

    # Both return an alias; the second multiplies the value by 10:
    e1 = TestFactory7(tag.alias, tag.pointer, 9)
    e2 = PythFactory7(tag.alias, tag.pointer, 10)
    assert e1.get() == 9
    assert e2.get() == 200
    assert e1.has_alias()
    assert e2.has_alias()

    f1 = TestFactory7(tag.shared_ptr, tag.base, 11)
    f2 = PythFactory7(tag.shared_ptr, tag.base, 12)
    assert f1.get() == 11
    assert f2.get() == 112
    assert not f1.has_alias()
    assert f2.has_alias()

    g1 = TestFactory7(tag.shared_ptr, tag.invalid_base, 13)
    assert g1.get() == 13
    assert not g1.has_alias()
    with pytest.raises(TypeError) as excinfo:
        PythFactory7(tag.shared_ptr, tag.invalid_base, 14)
    assert (str(excinfo.value) ==
            "pybind11::init(): construction failed: returned holder-wrapped instance is not an "
            "alias instance")

    assert [i.alive() for i in cstats] == [13, 7]
    assert ConstructorStats.detail_reg_inst() == n_inst + 13

    del a1, a2, b1, d1, e1, e2
    assert [i.alive() for i in cstats] == [7, 4]
    assert ConstructorStats.detail_reg_inst() == n_inst + 7
    del b2, c1, c2, d2, f1, f2, g1
    assert [i.alive() for i in cstats] == [0, 0]
    assert ConstructorStats.detail_reg_inst() == n_inst

    assert [i.values() for i in cstats] == [
        ["1", "2", "3", "4", "5", "6", "7", "8", "9", "100", "11", "12", "13", "14"],
        ["2", "4", "6", "8", "9", "100", "12"]
    ]
